### PR TITLE
fix(771): Adding mTLS support in QUIC

### LINKED ISSUE #771: Mutual authentication (mTLS) does not work with quic transport
### Describe the bug

I followed the instructions here https://zenoh.io/docs/manual/tls/#mutual-authentication-mtls to configure mTLS authentication with the quic transport.

My router is configured like this
```
{
  mode : "router",
  listen : {
    endpoints : [
      "quic/socket.hostname:7449",
    ]
  },
  transport: {
    link: {
      tls : {
        root_ca_certificate:"/path/to/chargercmd/client/minica.pem",
        client_auth : true,
        server_private_key:"/path/to/privkey.pem",
        server_certificate:"/path/to/fullchain.pem"
      }
    },
  },
}
```

With the server configured as above, I tried to publish using z_put.  I did not specific a client certificate in the z_put configuration.  I expected the publish to fail.

However, the publish was successful. I think the server ignores the `client_auth: true`  setting when using the quic protocol.

This fails as expected with transport set to 'tls'

### To reproduce

1. Start router with quic protocol and mTLS enabled as described in https://zenoh.io/docs/manual/tls/#mutual-authentication-mtls
2. Publish to a topic using a client configuration without specifying the client certificates.
3. Publish suceeds.  It is expected that the publish will fail as the client authentication is not performed.

### System info

- Platform : Ubuntu 22.04 on X86_64

### PR DESCRIPTION
**Pull request by gabrik**
This PR contains

- [x] Some refactoring for shared method between TLS and QUIC transport
- [x] mTLS support for QUIC
- [x] tests

Closes #771

### CODE REVIEW

**Comment by gabrik** [comment]
This may be blocked by: https://github.com/quinn-rs/quinn/issues/1737

**Comment by gabrik** [comment]
I was using `rustls` 0.23 that introduced this: [CryptoProvider](https://docs.rs/rustls/latest/rustls/crypto/struct.CryptoProvider.html#using-the-per-process-default-cryptoprovider) that **must** be initialized early in any process using TLS, thus tests where failing.

I did a rollback to 0.22 (the same version used in `main`) to make the test pass and check if the logic is correct.

File: Cargo.toml [hunk 1]
```diff
@@ -122,7 +122,7 @@ petgraph = "0.6.3"
 pnet = "0.34"
 pnet_datalink = "0.34"
 proc-macro2 = "1.0.51"
-quinn = "0.10.1"
+quinn = {version = "0.10.1"}
```
Comment (Mallets) [inline Cargo.toml:125]:
```suggestion
quinn = "0.10.1"
```

File: Cargo.toml [hunk 2]
```diff
@@ -132,7 +132,7 @@ ron = "0.8.1"
 ringbuffer-spsc = "0.1.9"
 rsa = "0.9"
 rustc_version = "0.4.0"
-rustls = "0.22.2"
+rustls = {version = "0.22.2"}
```
Comment (Mallets) [inline Cargo.toml:135]:
```suggestion
rustls = "0.22.2"
```

File: io/zenoh-links/zenoh-link-quic/Cargo.toml [hunk 3]
```diff
@@ -12,39 +12,46 @@
...
 [dependencies]
 async-trait = { workspace = true }
 base64 = { workspace = true }
 futures = { workspace = true }
-tracing = {workspace = true}
 quinn = { workspace = true }
 rustls-native-certs = { workspace = true }
-rustls-pemfile = { workspace = true }
+rustls-pemfile = { version = "1" }
+rustls-pki-types = { version = "1.0" }
```
Comment (Mallets) [inline io/zenoh-links/zenoh-link-quic/Cargo.toml:34]: Same comment as above.

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 4]
```diff
@@ -219,55 +217,12 @@ impl LinkManagerUnicastTrait for LinkManagerUnicastQuic {
...
-        } else {
-            client_crypto
-                .with_custom_certificate_verifier(Arc::new(WebPkiVerifierAnyServerName::new(
-                    root_cert_store,
-                )))
-                .with_no_client_auth()
-        };
+        let mut client_crypto = TlsClientConfig::new(&epconf)
+            .await
+            .map_err(|e| zerror!("Cannot create a new QUIC client on {addr}. {e}"))?;
```
Comment (Mallets) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs:222]:
```suggestion
            .map_err(|e| zerror!("Cannot create a new QUIC client on {addr}: {e}"))?;
```

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 5]
```diff
@@ -314,61 +271,14 @@ impl LinkManagerUnicastTrait for LinkManagerUnicastQuic {
...
         // Server config
-        let mut server_crypto = rustls::ServerConfig::builder()
-            .with_safe_defaults()
-            .with_no_client_auth()
-            .with_single_cert(certificates, private_key)?;
-        server_crypto.alpn_protocols = ALPN_QUIC_HTTP.iter().map(|&x| x.into()).collect();
-        let mut server_config = quinn::ServerConfig::with_crypto(Arc::new(server_crypto));
+        let mut server_crypto = TlsServerConfig::new(&epconf)
+            .await
+            .map_err(|e| zerror!("Cannot create a new QUIC listener on {addr}. {e}"))?;
```
Comment (Mallets) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs:277]:
```suggestion
            .map_err(|e| zerror!("Cannot create a new QUIC listener on {addr}: {e}"))?;
```

File: io/zenoh-transport/tests/endpoints.rs [hunk 6]
```diff
@@ -334,7 +334,7 @@ AXVFFIgCSluyrolaD6CWD9MqOex4YOfJR2bNxI7lFvuK4AwjyUJzT1U1HXib17mM
 #[cfg(feature = "transport_quic")]
 #[tokio::test(flavor = "multi_thread", worker_threads = 4)]
 async fn endpoint_quic() {
-    use zenoh_link::quic::config::*;
+    use zenoh_link::tls::config::*;
```
Comment (Mallets) [inline io/zenoh-transport/tests/endpoints.rs:337]: Here we have a conflict for `features`. This test runs for `#[cfg(feature = "transport_quic")]` bug it actually requires #[cfg(feature = "transport_tls")] as well for `use zenoh_link::tls::config::*;`. This means we can never compile `transport_quic` without activating `transport_tls` feature.

File: io/zenoh-links/zenoh-link-quic/src/lib.rs [hunk 7]
```diff
@@ -155,27 +76,22 @@ pub mod config {
 
     pub const TLS_SERVER_PRIVATE_KEY_FILE: &str = "server_private_key_file";
     pub const TLS_SERVER_PRIVATE_KEY_RAW: &str = "server_private_key_raw";
-    pub const TLS_SERVER_PRIVATE_KEY_BASE64: &str = "server_private_key_base64";
+    pub const TLS_SERVER_PRIVATE_KEY_BASE_64: &str = "server_private_key_base64";
```
Comment (Mallets) [inline io/zenoh-links/zenoh-link-quic/src/lib.rs:79]:
```suggestion
    pub const TLS_SERVER_PRIVATE_KEY_BASE64: &str = "server_private_key_base64";
```

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 8]
```diff
@@ -12,29 +12,27 @@
...
 use std::fmt;
-use std::io::BufReader;
 use std::net::IpAddr;
 use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr};
 use std::sync::Arc;
 use std::time::Duration;
 use tokio::sync::Mutex as AsyncMutex;
 use tokio_util::sync::CancellationToken;
 use zenoh_core::zasynclock;
+
```
Comment (Mallets) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs:29]:
```suggestion
```

File: io/zenoh-links/zenoh-link-quic/src/utils.rs [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+        match (c.server_private_key(), c.server_private_key_base64()) {
+            (Some(_), Some(_)) => {
+                bail!("Only one between 'server_private_key' and 'server_private_key_base64' can be present!")
+            }
+            (Some(server_private_key), None) => {
+                ps.push((TLS_SERVER_PRIVATE_KEY_FILE, server_private_key));
+            }
+            (None, Some(server_private_key)) => {
+                ps.push((
+                    TLS_SERVER_PRIVATE_KEY_BASE_64,
```
Comment (Mallets) [inline io/zenoh-links/zenoh-link-quic/src/utils.rs:72]:
```suggestion
                    TLS_SERVER_PRIVATE_KEY_BASE64,
```

File: io/zenoh-links/zenoh-link-quic/src/utils.rs [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+        };
+        Ok(TlsServerConfig { server_config: sc })
+    }
+
+    async fn load_tls_private_key(config: &Config<'_>) -> ZResult<Vec<u8>> {
+        load_tls_key(
+            config,
+            TLS_SERVER_PRIVATE_KEY_RAW,
+            TLS_SERVER_PRIVATE_KEY_FILE,
+            TLS_SERVER_PRIVATE_KEY_BASE_64,
```
Comment (Mallets) [inline io/zenoh-links/zenoh-link-quic/src/utils.rs:229]:
```suggestion
            TLS_SERVER_PRIVATE_KEY_BASE64,
```

**Pull request merged**